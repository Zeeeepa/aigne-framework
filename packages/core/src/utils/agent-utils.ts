import type { AgentHooks } from "../agents/agent.ts";

const priorities: NonNullable<AgentHooks["priority"]>[] = ["high", "medium", "low"];

export function sortHooks(hooks: AgentHooks[]): AgentHooks[] {
  return hooks
    .slice(0)
    .sort(({ priority: a = "low" }, { priority: b = "low" }) =>
      a === b ? 0 : priorities.indexOf(a) - priorities.indexOf(b),
    );
}
