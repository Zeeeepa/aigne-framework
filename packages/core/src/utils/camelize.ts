export type CamelCase<S extends string> = S extends `${infer P1}_${infer P2}${infer P3}`
  ? `${P1}${Uppercase<P2>}${CamelCase<P3>}`
  : S;

export type CamelizeObject<T, S = false> = {
  [K in keyof T as Uncapitalize<CamelCase<string & K>>]: T[K] extends Date
    ? T[K]
    : T[K] extends RegExp
      ? T[K]
      : T[K] extends Array<infer U>
        ? U extends object | undefined
          ? Array<CamelizeObject<U>>
          : T[K]
        : T[K] extends object | undefined
          ? S extends true
            ? T[K]
            : CamelizeObject<T[K]>
          : T[K];
};

export type Camelize<T, S = false> = T extends Array<infer U>
  ? Array<CamelizeObject<U, S>>
  : CamelizeObject<T, S>;

export function camelize<T, S extends boolean = false>(
  obj: T,
  shallow?: S,
): T extends string ? string : Camelize<T, S> {
  return typeof obj === "string" ? (camelCase(obj) as any) : walk(obj, shallow, camelCase);
}

export function snakelize<T, S extends boolean = false>(
  obj: T,
  shallow?: S,
): T extends string ? string : any {
  return typeof obj === "string" ? (snakeCase(obj) as any) : walk(obj, shallow, snakeCase);
}

function camelCase(key: string): string {
  key = key.replace(/[_.-](\w|$)/g, (_, char) => char.toUpperCase());
  key = key.charAt(0).toLowerCase() + key.slice(1);
  return key;
}

function snakeCase(key: string): string {
  return key.replace(/([a-z])([A-Z])/g, "$1_$2").toLowerCase();
}

function walk(obj: any, shallow = false, transform: (key: string) => string): any {
  if (!obj || typeof obj !== "object") return obj;
  if (obj instanceof Date || obj instanceof RegExp) return obj;

  if (Array.isArray(obj))
    return obj.map((v) => {
      if (!shallow) {
        return walk(v, shallow, transform);
      }
      if (typeof v === "object") return walk(v, shallow, transform);
      return v;
    });

  return Object.keys(obj).reduce(
    (res, key) => {
      const newKey = transform(key);
      res[newKey] = shallow ? obj[key] : walk(obj[key], shallow, transform);
      return res;
    },
    {} as Record<string, any>,
  );
}
