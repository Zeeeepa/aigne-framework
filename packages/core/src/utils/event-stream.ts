import { createParser } from "eventsource-parser";
import { produce } from "immer";
import {
  type AgentResponseChunk,
  type AgentResponseProgress,
  type AgentResponseStream,
  isAgentResponseDelta,
  isAgentResponseProgress,
  type Message,
} from "../agents/agent.js";
import type { Context, ContextEventMap } from "../aigne/context.js";
import { tryOrThrow } from "./type-utils.js";
import type { Listener } from "./typed-event-emitter.js";

export class EventStreamParser<T> extends TransformStream<string, T | Error> {
  constructor() {
    let parser: ReturnType<typeof createParser> | undefined;

    super({
      start(controller) {
        parser = createParser({
          onEvent: (event) => {
            const json = tryOrThrow(
              () => JSON.parse(event.data) as T | { message: string },
              (e) => {
                controller.enqueue(
                  new Error(`Parse response chunk json error: ${e.message} ${event.data}`),
                );
              },
            );
            if (json) {
              switch (event.event) {
                case "error":
                  controller.enqueue(new Error((json as { message: string }).message));
                  break;
                default: {
                  if (!event.event) controller.enqueue(json as T);
                  else console.warn(`Unknown event type: ${event.event}`, event.data);
                }
              }
            }
          },
        });
      },
      transform(chunk) {
        parser?.feed(chunk);
      },
    });
  }
}

export class AgentResponseStreamParser<O extends Message> extends TransformStream<
  AgentResponseChunk<O> | Error,
  AgentResponseChunk<O>
> {
  private json: O = {} as O;

  constructor() {
    super({
      transform: (chunk, controller) => {
        if (chunk instanceof Error) {
          controller.error(chunk);
          controller.terminate();
          return;
        }

        if (isAgentResponseDelta(chunk)) {
          this.json = produce(this.json, (draft) => {
            if (chunk.delta.json) Object.assign(draft, chunk.delta.json);

            if (chunk.delta.text) {
              for (const [key, text] of Object.entries(chunk.delta.text)) {
                const original = draft[key] as string | undefined;
                const t = (original || "") + (text || "");
                if (t) Object.assign(draft, { [key]: t });
              }
            }
          });

          controller.enqueue({
            ...chunk,
            delta: {
              ...chunk.delta,
              json: this.json,
            },
          });
        } else if (isAgentResponseProgress(chunk)) {
          if (chunk.progress.event === "agentFailed") {
            const { name, message } = chunk.progress.error;
            chunk.progress.error = new Error(message);
            chunk.progress.error.name = name;
          }
          controller.enqueue(chunk);
        }
      },
    });
  }
}

export class AgentResponseStreamSSE<O extends Message> extends ReadableStream<string> {
  constructor(stream: AgentResponseStream<O>) {
    let reader: ReadableStreamDefaultReader<AgentResponseChunk<O>> | undefined;

    super({
      async pull(controller) {
        reader ??= stream.getReader();
        try {
          const { value, done } = await reader.read();
          if (done) {
            controller.close();
            return;
          }

          if (isAgentResponseProgress(value)) {
            if (value.progress.event === "agentFailed") {
              value.progress.error = {
                name: value.progress.error.name,
                message: value.progress.error.message,
              };
            }
          }

          controller.enqueue(`data: ${JSON.stringify(value)}\n\n`);
        } catch (error) {
          controller.enqueue(
            `event: error\ndata: ${JSON.stringify({ message: error.message })}\n\n`,
          );
          controller.close();
        }
      },
    });
  }
}

export class AgentResponseProgressStream extends ReadableStream<AgentResponseProgress> {
  constructor(context: Context) {
    super({
      async start(controller) {
        const writeEvent = (
          eventName: keyof ContextEventMap,
          event: ContextEventMap[typeof eventName][0],
        ) => {
          const progress = {
            ...event,
            event: eventName,
            agent: { name: event.agent.name },
          } as AgentResponseProgress["progress"];

          controller.enqueue({ progress });
        };

        const close = () => {
          context.off("agentStarted", onAgentStarted);
          context.off("agentSucceed", onAgentSucceed);
          context.off("agentFailed", onAgentFailed);
          controller.close();
        };

        const onAgentStarted: Listener<"agentStarted", ContextEventMap> = async (event) => {
          const taskTitle = await event.agent.renderTaskTitle(event.input);

          writeEvent("agentStarted", { ...event, taskTitle });
        };
        const onAgentSucceed: Listener<"agentSucceed", ContextEventMap> = (event) => {
          writeEvent("agentSucceed", event);
          if (event.contextId === context.id) {
            close();
          }
        };
        const onAgentFailed: Listener<"agentFailed", ContextEventMap> = (event) => {
          writeEvent("agentFailed", event);
          if (event.contextId === context.id) {
            close();
          }
        };

        context.on("agentStarted", onAgentStarted);
        context.on("agentSucceed", onAgentSucceed);
        context.on("agentFailed", onAgentFailed);
      },
    });
  }
}
