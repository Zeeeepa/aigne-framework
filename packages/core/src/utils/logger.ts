import { nodejs } from "@aigne/platform-helpers/nodejs/index.js";
import debug from "debug";

export enum LogLevel {
  ERROR = "error",
  WARN = "warn",
  INFO = "info",
  DEBUG = "debug",
}

const levels = Object.values(LogLevel);

export function getLevelFromEnv(ns: string): LogLevel | undefined {
  for (const level of levels.toReversed()) {
    if (debug.enabled(`${ns}:${level}`)) {
      return level;
    }
  }
}

export class Logger {
  constructor(public options: { level: LogLevel; ns: string }) {
    this.debugLogger = debug(`${options.ns}:${LogLevel.DEBUG}`);
    this.infoLogger = debug(`${options.ns}:${LogLevel.INFO}`);
    this.warnLogger = debug(`${options.ns}:${LogLevel.WARN}`);
    this.errorLogger = debug(`${options.ns}:${LogLevel.ERROR}`);

    this.level = getLevelFromEnv(options.ns) || options.level;

    for (const logger of [this.debugLogger, this.infoLogger, this.warnLogger]) {
      // @ts-ignore
      logger.useColors = nodejs.isStdoutATTY;
      logger.enabled = true;
      logger.log = (...args: unknown[]) => this.logMessage(...args);
    }

    this.errorLogger.log = (...args: unknown[]) => this.logError(...args);
    // @ts-ignore
    this.errorLogger.useColors = nodejs.isStderrATTY;
    this.errorLogger.enabled = true;
  }

  level: LogLevel;

  private debugLogger: debug.Debugger;

  private infoLogger: debug.Debugger;

  private warnLogger: debug.Debugger;

  private errorLogger: debug.Debugger;

  logMessage = console.log;

  logError = console.error;

  debug(message: string, ...args: unknown[]) {
    if (this.enabled(LogLevel.DEBUG)) {
      this.debugLogger(message, ...args);
    }
  }

  info(message: string, ...args: unknown[]) {
    if (this.enabled(LogLevel.INFO)) {
      this.infoLogger(message, ...args);
    }
  }

  warn(message: string, ...args: unknown[]) {
    if (this.enabled(LogLevel.WARN)) {
      this.warnLogger(message, ...args);
    }
  }

  error(message: string, ...args: unknown[]) {
    if (this.enabled(LogLevel.ERROR)) {
      this.errorLogger(message, ...args);
    }
  }

  enabled(level: LogLevel) {
    return levels.indexOf(this.level) >= levels.indexOf(level);
  }
}

export const logger = new Logger({
  ns: "aigne:core",
  level: LogLevel.INFO,
});
