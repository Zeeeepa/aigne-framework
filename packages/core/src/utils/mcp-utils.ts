import { type JsonSchema, jsonSchemaToZod } from "@aigne/json-schema-to-zod";
import { UriTemplate } from "@modelcontextprotocol/sdk/shared/uriTemplate.js";
import {
  CallToolResultSchema,
  GetPromptResultSchema,
  type ListPromptsResult,
  type ListResourcesResult,
  type ListResourceTemplatesResult,
  type ListToolsResult,
  ReadResourceResultSchema,
} from "@modelcontextprotocol/sdk/types.js";
import { type ZodObject, type ZodType, z } from "zod";
import { type MCPBaseOptions, MCPPrompt, MCPResource, MCPTool } from "../agents/mcp-agent.js";
import { isEmpty } from "./type-utils.js";

export function toolFromMCPTool(tool: ListToolsResult["tools"][number], options: MCPBaseOptions) {
  return new MCPTool({
    ...options,
    name: tool.name,
    description: tool.description,
    inputSchema: isEmpty(tool.inputSchema.properties)
      ? z.object({})
      : jsonSchemaToZod<ZodObject<Record<string, ZodType>>>(tool.inputSchema as JsonSchema),
    outputSchema: CallToolResultSchema as ZodType<any>,
  });
}

export function promptFromMCPPrompt(
  prompt: ListPromptsResult["prompts"][number],
  options: MCPBaseOptions,
) {
  return new MCPPrompt({
    ...options,
    name: prompt.name,
    description: prompt.description,
    inputSchema: jsonSchemaToZod<ZodObject<Record<string, ZodType>>>({
      type: "object",
      properties:
        prompt.arguments &&
        Object.fromEntries(
          prompt.arguments.map((i) => [i.name, { type: "string", description: i.description }]),
        ),
      required: prompt.arguments?.filter((i) => i.required).map((i) => i.name),
    }),
    outputSchema: GetPromptResultSchema,
  });
}

export function resourceFromMCPResource(
  resource:
    | ListResourcesResult["resources"][number]
    | ListResourceTemplatesResult["resourceTemplates"][number],
  options: MCPBaseOptions,
) {
  const [uri, variables] = isResourceTemplate(resource)
    ? [resource.uriTemplate, new UriTemplate(resource.uriTemplate).variableNames]
    : [resource.uri, []];

  return new MCPResource({
    ...options,
    name: resource.name,
    uri,
    description: resource.description,
    inputSchema: z.object(Object.fromEntries(variables.map((i) => [i, z.string()]))),
    outputSchema: ReadResourceResultSchema,
  });
}

function isResourceTemplate(
  resource:
    | ListResourcesResult["resources"][number]
    | ListResourceTemplatesResult["resourceTemplates"][number],
): resource is ListResourceTemplatesResult["resourceTemplates"][number] {
  return typeof resource.uriTemplate === "string";
}
