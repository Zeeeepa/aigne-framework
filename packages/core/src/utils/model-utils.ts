import type { ChatModelOutputUsage } from "../agents/chat-model.js";
import type { Nullish } from "./type-utils.js";

export function mergeUsage(...usages: Nullish<ChatModelOutputUsage>[]): ChatModelOutputUsage {
  return {
    inputTokens: usages.reduce((acc, usage) => (usage ? acc + usage.inputTokens : acc), 0),
    outputTokens: usages.reduce((acc, usage) => (usage ? acc + usage.outputTokens : acc), 0),
  };
}
