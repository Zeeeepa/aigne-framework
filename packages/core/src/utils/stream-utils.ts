import equal from "fast-deep-equal";
import {
  type AgentProcessAsyncGenerator,
  type AgentResponseChunk,
  type AgentResponseStream,
  isAgentResponseDelta,
  isEmptyChunk,
  type Message,
} from "../agents/agent.js";
import { isNonNullable, isRecord, omitBy, type PromiseOrValue } from "./type-utils.js";
import "./stream-polyfill.js";
import type { ReadableStreamDefaultReadResult } from "bun";

export function objectToAgentResponseStream<T extends Message>(json: T): AgentResponseStream<T> {
  if (!isRecord(json)) {
    throw new Error(
      `expect to return a record type such as {result: ...}, but got (${typeof json}): ${json}`,
    );
  }

  return new ReadableStream({
    pull(controller) {
      controller.enqueue({ delta: { json } });
      controller.close();
    },
  });
}

export function mergeAgentResponseChunk<T extends Message>(
  output: T,
  chunk: AgentResponseChunk<T>,
) {
  if (isAgentResponseDelta(chunk)) {
    if (chunk.delta.text) {
      for (const [key, text] of Object.entries(chunk.delta.text)) {
        const original = output[key] as string | undefined;
        const t = (original || "") + (text || "");
        if (t) Object.assign(output, { [key]: t });
      }
    }

    if (chunk.delta.json) {
      Object.assign(
        output,
        omitBy(chunk.delta.json, (v) => v === undefined),
      );
    }
  }

  return output;
}

export async function agentResponseStreamToObject<T extends Message>(
  stream: AgentResponseStream<T> | AgentProcessAsyncGenerator<T>,
): Promise<T> {
  const json: T = {} as T;

  if (stream instanceof ReadableStream) {
    for await (const value of stream) {
      mergeAgentResponseChunk(json, value);
    }
  } else {
    for (;;) {
      const chunk = await stream.next();
      if (chunk.value) {
        if (chunk.done) {
          Object.assign(json, chunk.value);
        } else {
          mergeAgentResponseChunk(json, chunk.value);
        }
      }
      if (chunk.done) break;
    }
  }

  return json;
}

export function asyncGeneratorToReadableStream<T extends Message>(
  generator: AgentProcessAsyncGenerator<T>,
): AgentResponseStream<T> {
  return new ReadableStream({
    async pull(controller) {
      try {
        const chunk = await generator.next();
        if (chunk.value) {
          if (chunk.done) {
            controller.enqueue({ delta: { json: chunk.value } });
          } else {
            controller.enqueue(chunk.value);
          }
        }

        if (chunk.done) {
          controller.close();
        }
      } catch (error) {
        controller.error(error);
      }
    },
  });
}

export function onAgentResponseStreamEnd<T extends Message>(
  stream: AgentResponseStream<T>,
  options?: {
    onChunk?: (
      chunk: AgentResponseChunk<T>,
    ) => PromiseOrValue<AgentResponseChunk<T> | undefined | void>;
    onResult?: (result: T) => PromiseOrValue<Partial<T> | undefined | void>;
    onError?: (error: Error) => PromiseOrValue<Error>;
  },
): AgentResponseStream<T> {
  const json: T = {} as T;
  const reader = stream.getReader();

  return new ReadableStream({
    async pull(controller) {
      try {
        while (true) {
          const { value, done } = await reader.read();

          if (done) {
            const result = (await options?.onResult?.(json)) ?? json;

            if (result && !equal(result, json)) {
              let chunk: AgentResponseChunk<T> = { delta: { json: result } };
              chunk = (await options?.onChunk?.(chunk)) ?? chunk;
              controller.enqueue(chunk);
            }

            controller.close();

            return;
          }

          mergeAgentResponseChunk(json, value);

          const chunk = (await options?.onChunk?.(value)) ?? value;
          if (!isEmptyChunk(chunk)) {
            controller.enqueue(chunk);
            break;
          }
        }
      } catch (error) {
        controller.error((await options?.onError?.(error)) ?? error);
      }
    },
  });
}

export function isAsyncGenerator<T extends AsyncGenerator>(
  value: AsyncGenerator | unknown,
): value is T {
  return typeof value === "object" && value !== null && Symbol.asyncIterator in value;
}

export async function* arrayToAgentProcessAsyncGenerator<T extends Message>(
  chunks: (AgentResponseChunk<T> | Error)[],
  result?: Partial<T>,
): AgentProcessAsyncGenerator<T> {
  for (const chunk of chunks) {
    if (chunk instanceof Error) throw chunk;

    yield chunk;
  }
  if (result !== undefined) return result;
}

export function arrayToReadableStream<T>(chunks: (T | Error)[]): ReadableStream<T> {
  const list = [...chunks];

  return new ReadableStream({
    pull(controller) {
      const item = list.shift();
      if (!item) {
        controller.close();
        return;
      }

      if (item instanceof Error) {
        controller.error(item);
        return;
      }

      controller.enqueue(item);
    },
  });
}

export async function readableStreamToArray<T>(
  stream: ReadableStream<T>,
  options: { catchError: true },
): Promise<(T | Error)[]>;
export async function readableStreamToArray<T>(
  stream: ReadableStream<T>,
  options?: { catchError?: boolean },
): Promise<T[]>;
export async function readableStreamToArray<T>(
  stream: ReadableStream<T>,
  options?: { catchError?: boolean },
): Promise<(T | Error)[]> {
  const result: T[] = [];
  try {
    for await (const value of stream) {
      result.push(value);
    }
  } catch (error) {
    if (!options?.catchError) throw error;

    result.push(error);
  }
  return result;
}

export function stringToAgentResponseStream(
  str: string,
  key: "text" | string = "text",
): AgentResponseStream<Message> {
  const segmenter = new Intl.Segmenter(undefined, { granularity: "word" });
  const segments = segmenter.segment(str);

  return arrayToReadableStream(
    Array.from(segments).map((segment) => ({ delta: { text: { [key]: segment.segment } } })),
  );
}

export function toReadableStream(stream: NodeJS.ReadStream) {
  return new ReadableStream<Uint8Array>({
    start(controller) {
      const onData = (chunk: Buffer) => {
        controller.enqueue(new Uint8Array(chunk));
      };

      const onEnd = () => {
        cleanup();
        controller.close();
      };

      const onError = (err: Error) => {
        cleanup();
        controller.error(err);
      };

      function cleanup() {
        stream.off("data", onData);
        stream.off("end", onEnd);
        stream.off("error", onError);
      }

      stream.on("data", onData);
      stream.on("end", onEnd);
      stream.on("error", onError);
    },
  });
}

export async function readAllString(stream: NodeJS.ReadStream | ReadableStream): Promise<string> {
  return (
    await readableStreamToArray(
      (stream instanceof ReadableStream ? stream : toReadableStream(stream)).pipeThrough(
        new TextDecoderStream(),
      ),
    )
  ).join("");
}

export function mergeReadableStreams<T1, T2>(
  s1: ReadableStream<T1>,
  s2: ReadableStream<T2>,
): ReadableStream<T1 | T2>;
export function mergeReadableStreams(
  ...streams: (ReadableStream<any> | undefined)[]
): ReadableStream<any>;
export function mergeReadableStreams(
  ...streams: (ReadableStream<any> | undefined)[]
): ReadableStream<any> {
  type Reader = {
    reader: ReadableStreamDefaultReader;
    reading?: Promise<{
      result: ReadableStreamDefaultReadResult<any>;
      item: any;
    }>;
  };

  let readers: Reader[] | undefined;

  return new ReadableStream({
    async pull(controller) {
      try {
        readers ??= streams.filter(isNonNullable).map((s) => ({ reader: s.getReader(), data: [] }));

        while (readers.length) {
          const chunk: Awaited<NonNullable<Reader["reading"]>> = await Promise.race(
            readers.map((i) => {
              i.reading ??= i.reader.read().then((result) => ({ result, item: i }));
              return i.reading;
            }),
          );

          if (chunk.result.value) {
            controller.enqueue(chunk.result.value);
            chunk.item.reading = undefined;
            return;
          }

          if (chunk.result.done) {
            readers = readers.filter((i) => i !== chunk.item);
          }
        }

        controller.close();
      } catch (error) {
        controller.error(error);
        if (readers) for (const item of readers) item.reader.releaseLock();
      }
    },
  });
}
