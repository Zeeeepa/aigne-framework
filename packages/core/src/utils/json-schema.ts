import type { ZodType, z } from "zod";
import { zodToJsonSchema } from "zod-to-json-schema";
import type { Message } from "../agents/agent.js";
import { logger } from "./logger.js";

export function outputSchemaToResponseFormatSchema(
  agentOutput: ZodType<Message>,
): Record<string, unknown> {
  return setAdditionPropertiesDeep(zodToJsonSchema(agentOutput), false);
}

function setAdditionPropertiesDeep<T>(schema: T, additionalProperties: boolean): T {
  if (Array.isArray(schema)) {
    return schema.map((s) => setAdditionPropertiesDeep(s, additionalProperties)) as T;
  }
  if (schema !== null && typeof schema === "object" && !Array.isArray(schema)) {
    return Object.entries(schema).reduce(
      (acc, [key, value]) => {
        acc[key] = setAdditionPropertiesDeep(value, additionalProperties);
        if (acc.type === "object") {
          acc.additionalProperties = additionalProperties;
        }
        return acc;
      },
      {} as Record<string, unknown>,
    ) as T;
  }
  return schema;
}

export function parseJSON(json: string) {
  try {
    return JSON.parse(json);
  } catch (error) {
    logger.error("Failed to parse JSON", { json, error });
    throw error;
  }
}

/**
 * Ensure that the union array has at least 1 item.
 * NOTE: the zod union requires at least 2 items (just type definition, not runtime behavior)
 * so we need to ensure that the union has at least 1 item.
 * @param union - The union array
 * @returns The union array with at least 1 item (but the type is at least 2 items for z.union)
 */
export function ensureZodUnionArray<T extends z.ZodType>(union: T[]): [T, T, ...T[]] {
  if (!(union.length >= 1)) {
    throw new Error(`Union must have at least 1 item, but got ${union.length}`);
  }
  return union as [T, T, ...T[]];
}
