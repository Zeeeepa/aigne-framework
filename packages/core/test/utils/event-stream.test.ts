import { expect, test } from "bun:test";
import type { AgentResponseChunk, Message } from "@aigne/core";
import {
  AgentResponseStreamParser,
  AgentResponseStreamSSE,
  EventStreamParser,
} from "@aigne/core/utils/event-stream.js";
import { arrayToReadableStream, readableStreamToArray } from "@aigne/core/utils/stream-utils";

test("EventStreamParser should enqueue an error for invalid json", async () => {
  const stream = arrayToReadableStream([
    `data: {"delta": {"text": {"text": "hello"}}}\n\n`,
    `data: {"delta": {"text": {"text": " world"}}}\n\n`,
    "data: invalid json string\n\n",
  ]).pipeThrough(new EventStreamParser());

  expect(readableStreamToArray(stream)).resolves.toMatchSnapshot();
});

test("EventStreamParser should enqueue an error for error event", async () => {
  const stream = arrayToReadableStream([
    `data: {"delta": {"text": {"text": "hello"}}}\n\n`,
    `data: {"delta": {"text": {"text": " world"}}}\n\n`,
    `event: error\ndata: {"message": "test error"}\n\n`,
  ]).pipeThrough(new EventStreamParser());

  expect(readableStreamToArray(stream)).resolves.toMatchSnapshot();
});

test("AgentResponseStreamParser should raise error from upstream", async () => {
  const stream = arrayToReadableStream<AgentResponseChunk<Message>>([
    { delta: { text: { text: "hello" } } },
    { delta: { text: { text: " world" } } },
    new Error("should raise error"),
    { delta: { text: { text: " never" } } },
  ]).pipeThrough(new AgentResponseStreamParser());

  expect(readableStreamToArray(stream)).rejects.toThrow("should raise error");
});

test("AgentResponseStreamSSE should convert AgentResponseStream to SSE format", async () => {
  const stream = new AgentResponseStreamSSE(
    arrayToReadableStream([
      { delta: { text: { text: "hello" } } },
      { delta: { text: { text: " world" } } },
    ]),
  );

  expect(readableStreamToArray(stream)).resolves.toEqual([
    `data: {"delta":{"text":{"text":"hello"}}}\n\n`,
    `data: {"delta":{"text":{"text":" world"}}}\n\n`,
  ]);
});
