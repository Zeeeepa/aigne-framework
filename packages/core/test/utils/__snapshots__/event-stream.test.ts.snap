// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`EventStreamParser should enqueue an error for invalid json 1`] = `
[
  {
    "delta": {
      "text": {
        "text": "hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " world",
      },
    },
  },
  [Error: Parse response chunk json error: JSON Parse error: Unexpected identifier "invalid" invalid json string],
]
`;

exports[`EventStreamParser should enqueue an error for error event 1`] = `
[
  {
    "delta": {
      "text": {
        "text": "hello",
      },
    },
  },
  {
    "delta": {
      "text": {
        "text": " world",
      },
    },
  },
  [Error: test error],
]
`;
